Time Series Annotation Layers

graceguo-supercat opened the pull request:
fixes https://github.com/apache/incubator-superset/issues/3502
![vudfmdwhhs](https://user-images.githubusercontent.com/27990562/30837001-ad00b3ca-a216-11e7-9e24-f65f85658158.gif)

mistercrunch committed:
Adding annotations to backend

mistercrunch committed:
Auto fetching Annotations on the backend

mistercrunch committed:
Closing the loop

mistercrunch committed:
Adding missing files

Commit:
annotation layers UI

for https://github.com/apache/incubator-superset/issues/3502

mistercrunch commented on superset/assets/javascripts/explore/components/controls/SelectAsyncControl.jsx:
```diff
@@ -0,0 +1,62 @@
+class SelectAsyncControl extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Many of the other controls to not maintain internal state and rely solely on the `value` prop and `onChange` to act as a controlled component. There's nothing bad with maintaining the state here but it adds a bid of complexity that is not needed.

graceguo-supercat replied:
fixed.

mistercrunch commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -476,6 +477,71 @@ function nvd3Vis(slice, payload) {
+          .attr('height', 10)
+          .attr('patternTransform', 'rotate(45 50 50)')
+          .append('line')
+          .attr('stroke', '#00A699')
```
Looked like this color is `@brand-primary`, any way we can use a CSS class instead of hard-coding the color here?

mistercrunch commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -476,6 +477,71 @@ function nvd3Vis(slice, payload) {
+          .attr('fill', 'url(#diagonal)')
+          .attr('fill-opacity', 0.1)
+          .attr('stroke-width', 1)
+          .attr('stroke', '#00A699')
```
same here, let's not hard code the color

mistercrunch commented on superset/migrations/versions/f959a6652acd_.py:
```diff
@@ -0,0 +1,22 @@
+"""empty message
```
I'm surprised to see 2 different merge migrations. Legit or mistake?

mistercrunch commented on superset/models/core.py:
```diff
@@ -25,7 +25,7 @@
 from sqlalchemy import (
     Column, Integer, String, ForeignKey, Text, Boolean,
-    DateTime, Date, Table,
+    DateTime, Date, Table, Index,
```
`Index` is unused, I think I left that dirt behind in my commit...

mistercrunch commented on superset/viz.py:
```diff
@@ -111,6 +112,7 @@ def get_extra_filters(self):
 
     def query_obj(self):
         """Building a query object"""
+        print("running query_obj============================")
```
Let's remove this line

mistercrunch commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -476,6 +477,71 @@ function nvd3Vis(slice, payload) {
+          .attr('class', 'd3-tip')
+          .direction('n')
+          .offset([-5, 0])
+          .html(d => (d && d.layer ? d.layer : ''));
```
I think we should surface more here, the layer name, the short and the long description.

mistercrunch commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -476,6 +477,71 @@ function nvd3Vis(slice, payload) {
+          })
+          .attr('height', hh)
+          .attr('fill', 'url(#diagonal)')
+          .attr('fill-opacity', 0.1)
```
note: I think the annotations look good on the gif, we can talk to Eli & Chris to fine tune their look if needed.

mistercrunch replied:
class could be called `stroke-primary`

mistercrunch reviewed (changes requested):
A few minor things, looks great!

Commit:
a few fixes per code review.

- add annotation input sanity check before add and before update.
- make SelectAsyncControl component statelesis, and generic
- add annotation description in d3 tool tip
- use less variable to replace hard-coded color

fabianmenges commented on superset/viz.py:
```diff
@@ -57,6 +57,7 @@ def __init__(self, datasource, form_data):
             'token', 'token_' + uuid.uuid4().hex[:8])
         self.metrics = self.form_data.get('metrics') or []
         self.groupby = self.form_data.get('groupby') or []
+        self.annotation_layers = []
```
I don't think annotations should be part the viz object. For most viz types that I can think of annotations don't really make sense. If you make annotations accessible from a json endpoint they can be loaded client side. This would also make the responsiveness be better, annotations could be loaded async independently of the data.

mistercrunch replied:
Here we decided to package the annotations from the backend along with the data payload. The visualizations' render method can then assume that all of the data + annotations + context is present when rendering. Note that the annotation query is fully indexed and should take milliseconds assuming a small-ish number of annotations for the time range.

I feel like if we wanted to go async on the annotations, we'd need to change the visualization's API to define 2 different methods and somehow guarantee to call the annotation one second. It makes things much more complicated on that side of things.

fabianmenges replied:
https://github.com/apache/incubator-superset/pull/3518/commits/34ec8b35a8eca6a6cef070387aa1c19bd4a972a1#diff-a85def8afdc240ac598637a26a7cdbbc I probably clean this up a bit more, but this makes all http requests in parallel.

fabianmenges commented on superset/views/annotations.py:
```diff
@@ -0,0 +1,59 @@
+from .base import SupersetModelView, DeleteMixin
+
+
+class AnnotationModelView(SupersetModelView, DeleteMixin):  # noqa
```
How are annotations written to that table? If you cannot write them from within superset, why are you not thinking of them as just another datasource?

mistercrunch replied:
This feature ships with a CRUD that allows users to manually input/manage layers. FAB also exposes a CURD REST API to maintain those programmatically, though I'm thinking at Airbnb most likely people will write Airflow jobs talking to our MySQL instance to load annotations dynamically.

fabianmenges replied:
I don't think that this table should be part of the Superset DB then. The way Superset is currently designed is that it is agnostic (as long as there is a Druid/SQL connector) to where the data is stored. We have a clean separation of meta data (slice definition, configuration, users, ...) and the data that superset is operating on.
This table inverses this fundamentally and now we have to store presentation data directly inside of the superset db. Currently no time series data is stored inside of the superset database. I'm also worried that this will cause scalability issues. Now my superset meta database is going to have very different performance requirements with regard to querying time series.

If this table was only used for manually created data from a user interface you could make the argument that we are talking about a very limited amount of data which originated in superset and therefore should be stored there. But this seems not to be the case.

mistercrunch replied:
Yeah it's at the frontier of data/metadata, though to me annotations are closer to metadata.

mistercrunch replied:
Though I agree that it would be sweet to support external data for annotations (as in your PR), let's do both. I have some thoughts on that that I'll put on your PR. The TLDR is I think we should create a new `time_annotations` viz_type to enforce a certain set of guaranties, as opposed to a convention on top of the `table` viz_type.

fabianmenges commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -476,6 +477,78 @@ function nvd3Vis(slice, payload) {
+
+      // add annotation_layer
+      if (isTimeSeries && payload.annotations.length) {
+        const tip = d3tip()
```
Do these tooltips work when we are using "Rich Tooltips" (useInteractiveGuideline) How does that look?

fabianmenges commented on superset/migrations/versions/ddd6ebdd853b_annotations.py:
```diff
@@ -0,0 +1,56 @@
+        'annotation',
+        sa.Column('created_on', sa.DateTime(), nullable=True),
+        sa.Column('changed_on', sa.DateTime(), nullable=True),
+        sa.Column('id', sa.Integer(), nullable=False),
```
Bigint?

mistercrunch replied:
hoping we never go above 2,147,483,647 annotations, saving some bytes

fabianmenges replied:
If you have automated processes writing to this table (airflow) we should change this to BIGINT. The disk space you save is only relevant in cases when you have a lot of rows, which is also when you want to have BIGINTs as your IDs.

fabianmenges commented:
I just created a couple youtube videos of the functionality of my changeset,  #3518 
I'd be happy to incorporate your annotation table and the ability to create "rectangular" annotations and just add it as another type of annotation. Currently my change set incorporates "Events", "Formulas", "Time Series". I could create another type "Annotation" that would pull the annotations from the Annotation table in the Superset DB. I could also then provide a UI where you can add annotations inside of Superset for that Layer.

mistercrunch commented:
@fabianmenges going to comment on your PR to think holistically about annotations. We spoke about your approach internally and think it's very complementary and would love to integrate all this.

mistercrunch commented:
@fabianmenges we may need to talk to sync up about annotations, we all really  like the work that you've done and do want to support the stuff you added:
* slice-defined formulas
* slice-defined vertical lines
* slice-defined data points (X/Y) annotations using http://d3-annotation.susielu.com/
* data-based annotations, using a specific slice-types with the right constraints and guarantees, reusable across slices
* Superset-managed Annotation Layers, to be renamed "Time Annotation Layers" to clarify, these are highly re-usage across slices.

All this should be composable.

I'd like to break down 1-2-3 from 4 potentially, though if we can agree on the vision we could make your current PR evolve into 1-2-4 combo. That's a pretty massive feature set though.

fabianmenges commented:
Lets move this to the dev mailing list...

fabianmenges commented:
Go ahead and merge this in, I can integrate the UIs in my merge request. Lets have the design discussion in the mailing list.
